# Movies Tonight Part 3: SQL DML

In [93]:
%load_ext sql
%sql sqlite:///Movies.db

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [94]:
%%sql

PRAGMA table_info(DATASET);

 * sqlite:///Movies.db
Done.


[(0, 'TName', 'TEXT', 0, None, 0),
 (1, 'Location', 'TEXT', 0, None, 0),
 (2, 'Phone', 'TEXT', 0, None, 0),
 (3, 'MTitle', 'TEXT', 0, None, 0),
 (4, 'ShowTime', 'TEXT', 0, None, 0),
 (5, 'Rating', 'TEXT', 0, None, 0),
 (6, 'CCode', 'TEXT', 0, None, 0),
 (7, 'CName', 'TEXT', 0, None, 0)]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [95]:
%%sql

INSERT INTO MOVIE (MTitle,Rating)
    SELECT DISTINCT MTitle, Rating
        FROM DATASET;

 * sqlite:///Movies.db
23 rows affected.


[]

In [96]:
%%sql

INSERT INTO THEATER (TName,Phone,Location)
    SELECT DISTINCT TName, Phone, Location
        FROM DATASET;

 * sqlite:///Movies.db
9 rows affected.


[]

In [97]:
%%sql

INSERT INTO ARTISTS (Name)
    SELECT DISTINCT CName
        FROM DATASET;

 * sqlite:///Movies.db
153 rows affected.


[]

In [99]:
%%sql

INSERT INTO CREDIT (MID,AID,CCode)
SELECT DISTINCT MID, AID, CCode
FROM MOVIE 
JOIN DATASET USING (MTitle)
JOIN ARTISTS ON (DATASET.CName = ARTISTS.Name);

 * sqlite:///Movies.db
162 rows affected.


[]

In [105]:
%%sql

INSERT INTO SHOW (MID,TID,Show_Time)
SELECT DISTINCT MID, TID, ShowTime
FROM MOVIE
JOIN DATASET USING (MTitle)
JOIN THEATER ON (DATASET.Location=THEATER.Location);

 * sqlite:///Movies.db
131 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [106]:
%%sql

SELECT MTitle, TName, ShowTime, CCode, CName
FROM DATASET
GROUP BY MTitle, TName, ShowTime, CCode, CName
HAVING COUNT(*) > 1;

 * sqlite:///Movies.db
Done.


[('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Greg Kinnear'),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Hector Elizondo'),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Jon Seda'),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Laurie Metcalf'),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Maria Pitillo'),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Roscoe Lee Browne'),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Tim Conway'),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'D', 'Garry Marshall'),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Jon Lovitz'),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Mekhi Phifer'),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Tia Carrera'),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'and Louise Fletcher'),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'D', 'Hart Bochner')]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

In [107]:
%%sql

DROP TABLE DATASET;

 * sqlite:///Movies.db
Done.


[]

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".